### ISSUE TITLE
Incompletely parsed structs can lead to undefined behavior

### ISSUE THREAD

**Issue by mmarchini** [iovisor/bpftrace#537]
I don't have an example right now, but our clang_parser implementation skips more complex fields inside structs and set their size to zero, which means for any field after those undefined types we'll probably get the wrong offset for the field, leading to undefined behavior. We should flag incomplete structs and throw an error (or warn) when users use them.

**Comment by dalehamel** [comment]
I have run into this numerous times on more complex structs.

Most of what is skipped I believe has to do with preprocessor macros, which makes the current approach for calculating offsets invalid, as you don't know if the macro is defined or not, and I'm not sure how you could reasonable do so at runtime, even if you read the running kernel config.

There should be ways to actually access the data, such as #285 if you are able to determine the offset of the address yourself, but ultimately I think we need btf to give the actual type information for the running kernel, rather than try to piece it together through headers and macros.

**Comment by fbs** [comment]
Are there some concrete examples to test with? Haven't bumped into this lately, might have been fixed?

**Comment by dalehamel** [comment]
Ultimately the best solution is to use BTF for this I think. I can help to come up with a specific example, but this is fairly easy to bump into with dynamic kernel tracing where the headers have multiple `#ifdef` macros. Basically bpftrace has no way of knowing what the preprocessor ultimately resolved the struct fields to.

However, I noticed newer versions of the kernel offer a way to embed the headers in the kernel - I haven't checked if they have this issue, as conceivably they could be resolved based on what the compiler actually evaluated, though I haven't confirmed this.

The traditional way of looking at /usr/src/linux or /lib/modules paths to get headers is how I ran into this.

I think we can bump this out from the 0.9.4 release though, if it's moved to 0.9.5 i'll look into a test case that shows the issue, and try to more completely evaluated the solution space.
